### ISSUE TITLE
Asking for support when adding GX430t label printer: Layout must be valid JSON

### ISSUE THREAD

**Issue by chkohler** [miso-lims/miso-lims#2342]
Dear Miso-Team,

I'm asking for your support as I'm struggling with adding a GX430t Zebra label printer (MISO version 1.10.1).
By the end of April this year, my colleague Nicholas had E-Mail contact with Dillan Cooke and Morgan Taschuk about Zebra printers that MISO does support.

> MISO does support Zebra printers. It is currently possible, though difficult, to use any label size. The label layouts are stored in the database and can be modified via SQL. This isn’t simple however, and we plan to add an editor for modifying these via the UI in the future. The default layouts include these label models for the Zebra printers:
>
> JTT-183: 25.4mm x 12.7mm + 9.5mm circle
>
>JTT-7: 25.4mm x 12.7mm

Following the [documentation on how to add Barcode Label Printers](https://miso-lims.readthedocs.io/projects/docs/en/latest/user_manual/barcode_label_printers/), I end up with providing the information as shown below
![Screenshot 2020-08-27 at 13 53 46](https://user-images.githubusercontent.com/40832074/91439151-c9ebb080-e86c-11ea-939a-dc55d0ee0736.png)

When submitting, an error window pops up:
![Screenshot 2020-08-27 at 13 30 51](https://user-images.githubusercontent.com/40832074/91438452-a6743600-e86b-11ea-9071-9de3aec2575b.png)

I tried various combinations in filling the _Layout_ field (also with combinations of _JTT-183_) but failed permanently.
I simply don't know which information to provide (and most importantly in which format) ?!
 
I found a [thread on the Gitter chat](https://gitter.im/miso-lims/community) but this didn't help me either (but at least Zebra printers seem to work fine with MISO)
![Screenshot 2020-08-27 at 13 49 31](https://user-images.githubusercontent.com/40832074/91438844-4df16880-e86c-11ea-9535-70ee2a5dfba2.png)

Could you please help me out and provide the information I need to get our GX430t online?

Thank you very much for any help.

Cheers
Christian

**Comment by djcooke** [comment]
Hello,

The layout field in the Create Printer dialog must be JSON describing the label layout. This is very ugly and undocumented right now. We have plans to add an editor to make this much more friendly at some point. The layout we are using for JTT-183 is:

```
[{"contents":{"use":"ALIAS"},"element":"textblock","height":0.5,"lineLimit":8,"rowLimit":2,"x":10.7,"y":2.3},{"contents":{"use":"ALIAS"},"element":"textblock","height":0.8,"lineLimit":18,"rowLimit":2,"x":1,"y":1.375},{"contents":[{"use":"DATE"},{"use":"DESCRIPTION"}],"element":"textblock","height":0.8,"lineLimit":12,"rowLimit":2,"x":1,"y":3.375},{"contents":{"use":"ALIAS"},"element":"2dbarcode","height":0.1,"x":6.78,"y":4}]
```

I was just testing this out and there seems to be a bug with creating printers though. I will try to get a fix in for the next release. In the meantime, you can create the printer using this SQL:

```
INSERT INTO Printer (name, driver, backend, configuration, height, width, layout) VALUES
('Zebra-GTX430t', 'ZEBRA_12DPMM', 'CUPS', 25.4, 12.7, '[{"contents":{"use":"ALIAS"},"element":"textblock","height":0.5,"lineLimit":8,"rowLimit":2,"x":10.7,"y":2.3},{"contents":{"use":"ALIAS"},"element":"textblock","height":0.8,"lineLimit":18,"rowLimit":2,"x":1,"y":1.375},{"contents":[{"use":"DATE"},{"use":"DESCRIPTION"}],"element":"textblock","height":0.8,"lineLimit":12,"rowLimit":2,"x":1,"y":3.375},{"contents":{"use":"ALIAS"},"element":"2dbarcode","height":0.1,"x":6.78,"y":4}]');
```

Please let me know if that works for you.

**Comment by chkohler** [comment]
Dear Dillan,

thanks for your quick reply.
After running your SQL Statement, I got this error:

> ERROR 1136 (21S01): Column count doesn't match value count at row 1

Checking the column / value pairs of the proposed SQL Statement, it seems that the value for the _configuration_ column is missing. What value do I need to provide?

**Comment by djcooke** [comment]
Oops, sorry about that. The configuration column should just be `'{}'`:

```
INSERT INTO Printer (name, driver, backend, configuration, height, width, layout) VALUES
('Zebra-GTX430t', 'ZEBRA_12DPMM', 'CUPS', '{}', 25.4, 12.7, '[{"contents":{"use":"ALIAS"},"element":"textblock","height":0.5,"lineLimit":8,"rowLimit":2,"x":10.7,"y":2.3},{"contents":{"use":"ALIAS"},"element":"textblock","height":0.8,"lineLimit":18,"rowLimit":2,"x":1,"y":1.375},{"contents":[{"use":"DATE"},{"use":"DESCRIPTION"}],"element":"textblock","height":0.8,"lineLimit":12,"rowLimit":2,"x":1,"y":3.375},{"contents":{"use":"ALIAS"},"element":"2dbarcode","height":0.1,"x":6.78,"y":4}]');
```

**Comment by chkohler** [comment]
Yeah, the printer now shows up in the "Printer" section. 

![Screenshot 2020-09-01 at 06 05 44](https://user-images.githubusercontent.com/40832074/91793720-44278680-ec19-11ea-8821-9e9393b3d970.png)

AFAICS there is currently no option to edit printer properties (such as IP or Port), right? 

I also noticed that adding additional printers fail in case the previous input data was faulty:

![Screenshot 2020-09-01 at 06 08 29](https://user-images.githubusercontent.com/40832074/91793949-024b1000-ec1a-11ea-8170-5bb257678119.png)

I guess this behaviour has been fixed in (#2344) ?

I'm now running v1.11.0 and will hence wait for the v.12.x release

**BTW**: Our GX430t printer is assigned a stable IP as provided by DHCP. So what _Backend_ option would you suggest (`CUPS` or `RAW_TCP`) ?

Thanks for your help, Dillan.

**Comment by djcooke** [comment]
> AFAICS there is currently no option to edit printer properties (such as IP or Port), right?

Correct.

> I also noticed that adding additional printers fail in case the previous input data was faulty:
> I guess this behaviour has been fixed in (#2344) ?

Correct.

> Our GX430t printer is assigned a stable IP as provided by DHCP. So what Backend option would you suggest (CUPS or RAW_TCP) ?

If they both work for you, then whichever you prefer. If you are trying to create this via SQL, `RAW_TCP` does require a host and port to be configured. The `configuration` column should be JSON similar to this: `{"host": "192.168.0.10", "port": "9100"}`.

**Comment by djcooke** [comment]
Errors creating printers have been fixed, and there is now an editor in MISO for designing labels
